File: airbyte-commons-worker/src/test/java/io/airbyte/workers/general/ReplicationWorkerHelperTest.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.general;

import static io.airbyte.workers.test_utils.TestConfigHelpers.DESTINATION_IMAGE;
import static io.airbyte.workers.test_utils.TestConfigHelpers.SOURCE_IMAGE;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

import io.airbyte.api.client.AirbyteApiClient;
import io.airbyte.api.client.generated.ActorDefinitionVersionApi;
import io.airbyte.api.client.generated.DestinationApi;
import io.airbyte.api.client.generated.SourceApi;
import io.airbyte.api.client.model.generated.ResolveActorDefinitionVersionResponse;
import io.airbyte.commons.concurrency.VoidCallable;
import io.airbyte.commons.converters.ThreadedTimeTracker;
import io.airbyte.commons.json.Jsons;
import io.airbyte.config.ConfiguredAirbyteCatalog;
import io.airbyte.config.ConfiguredAirbyteStream;
import io.airbyte.config.MapperConfig;
import io.airbyte.config.State;
import io.airbyte.config.StreamDescriptor;
import io.airbyte.config.WorkerDestinationConfig;
import io.airbyte.config.adapters.AirbyteJsonRecordAdapter;
import io.airbyte.featureflag.FeatureFlagClient;
import io.airbyte.featureflag.TestClient;
import io.airbyte.mappers.application.RecordMapper;
import io.airbyte.mappers.transformations.DestinationCatalogGenerator;
import io.airbyte.persistence.job.models.ReplicationInput;
import io.airbyte.protocol.models.AirbyteAnalyticsTraceMessage;
import io.airbyte.protocol.models.AirbyteLogMessage;
import io.airbyte.protocol.models.AirbyteMessage;
import io.airbyte.protocol.models.AirbyteMessage.Type;
import io.airbyte.protocol.models.AirbyteRecordMessage;
import io.airbyte.protocol.models.AirbyteTraceMessage;
import io.airbyte.workers.context.ReplicationContext;
import io.airbyte.workers.context.ReplicationFeatureFlags;
import io.airbyte.workers.helper.StreamStatusCompletionTracker;
import io.airbyte.workers.internal.AirbyteDestination;
import io.airbyte.workers.internal.AirbyteMapper;
import io.airbyte.workers.internal.AirbyteSource;
import io.airbyte.workers.internal.AnalyticsMessageTracker;
import io.airbyte.workers.internal.FieldSelector;
import io.airbyte.workers.internal.bookkeeping.AirbyteMessageOrigin;
import io.airbyte.workers.internal.bookkeeping.AirbyteMessageTracker;
import io.airbyte.workers.internal.bookkeeping.SyncStatsTracker;
import io.airbyte.workers.internal.bookkeeping.events.ReplicationAirbyteMessageEventPublishingHelper;
import io.airbyte.workers.internal.bookkeeping.streamstatus.StreamStatusTracker;
import io.airbyte.workers.internal.bookkeeping.streamstatus.StreamStatusTrackerFactory;
import io.airbyte.workers.internal.syncpersistence.SyncPersistence;
import io.airbyte.workload.api.client.WorkloadApiClient;
import io.airbyte.workload.api.client.generated.WorkloadApi;
import java.io.IOException;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;

class ReplicationWorkerHelperTest {

  private ReplicationWorkerHelper replicationWorkerHelper;
  private AirbyteMapper mapper;
  private SyncStatsTracker syncStatsTracker;
  private AirbyteMessageTracker messageTracker;
  private SyncPersistence syncPersistence;
  private AnalyticsMessageTracker analyticsMessageTracker;
  private StreamStatusCompletionTracker streamStatusCompletionTracker;
  private StreamStatusTracker streamStatusTracker;
  private StreamStatusTrackerFactory streamStatusTrackerFactory;
  private WorkloadApiClient workloadApiClient;
  private AirbyteApiClient airbyteApiClient;
  private ActorDefinitionVersionApi actorDefinitionVersionApi;
  private ReplicationAirbyteMessageEventPublishingHelper replicationAirbyteMessageEventPublishingHelper;
  private RecordMapper recordMapper;
  private FeatureFlagClient featureFlagClient;
  private DestinationCatalogGenerator destinationCatalogGenerator;

  private final ReplicationContext replicationContext = new ReplicationContext(true, UUID.randomUUID(), UUID.randomUUID(), UUID.randomUUID(), 0L,
      1, UUID.randomUUID(), SOURCE_IMAGE, DESTINATION_IMAGE, UUID.randomUUID(), UUID.randomUUID());

  @BeforeEach
  void setUp() {
    mapper = mock(AirbyteMapper.class);
    syncStatsTracker = mock(SyncStatsTracker.class);
    syncPersistence = mock(SyncPersistence.class);
    messageTracker = mock(AirbyteMessageTracker.class);
    analyticsMessageTracker = mock(AnalyticsMessageTracker.class);
    streamStatusCompletionTracker = mock(StreamStatusCompletionTracker.class);
    replicationAirbyteMessageEventPublishingHelper = mock(ReplicationAirbyteMessageEventPublishingHelper.class);
    streamStatusTracker = mock(StreamStatusTracker.class);
    streamStatusTrackerFactory = mock(StreamStatusTrackerFactory.class);
    workloadApiClient = mock(WorkloadApiClient.class);
    airbyteApiClient = mock(AirbyteApiClient.class);
    when(messageTracker.getSyncStatsTracker()).thenReturn(syncStatsTracker);
    when(workloadApiClient.getWorkloadApi()).thenReturn(mock(WorkloadApi.class));
    when(airbyteApiClient.getDestinationApi()).thenReturn(mock(DestinationApi.class));
    when(airbyteApiClient.getSourceApi()).thenReturn(mock(SourceApi.class));
    actorDefinitionVersionApi = mock(ActorDefinitionVersionApi.class);
    when(airbyteApiClient.getActorDefinitionVersionApi()).thenReturn(actorDefinitionVersionApi);
    when(streamStatusTrackerFactory.create(any())).thenReturn(streamStatusTracker);
    recordMapper = mock(RecordMapper.class);
    featureFlagClient = mock(TestClient.class);
    destinationCatalogGenerator = mock(DestinationCatalogGenerator.class);
    replicationWorkerHelper = spy(new ReplicationWorkerHelper(
        mock(FieldSelector.class),
        mapper,
        messageTracker,
        syncPersistence,
        replicationAirbyteMessageEventPublishingHelper,
        mock(ThreadedTimeTracker.class),
        mock(VoidCallable.class),
        workloadApiClient,
        analyticsMessageTracker,
        "workload-id",
        airbyteApiClient,
        streamStatusCompletionTracker,
        streamStatusTrackerFactory,
        recordMapper,
        featureFlagClient,
        destinationCatalogGenerator));
  }

  @AfterEach
  void tearDown() {
    Mockito.framework().clearInlineMocks();
  }

  @ParameterizedTest
  @ValueSource(booleans = {true, false})
  void testGetReplicationOutput(final boolean supportRefreshes) throws IOException {
    mockSupportRefreshes(supportRefreshes);
    // Need to pass in a replication context
    final ConfiguredAirbyteCatalog catalog = buildConfiguredAirbyteCatalog();
    when(destinationCatalogGenerator.generateDestinationCatalog(any()))
        .thenReturn(new DestinationCatalogGenerator.CatalogGenerationResult(catalog, Map.of()));
    replicationWorkerHelper.initialize(
        replicationContext,
        mock(ReplicationFeatureFlags.class),
        mock(Path.class),
        catalog,
        mock(State.class));
    verify(streamStatusCompletionTracker).startTracking(catalog, supportRefreshes);
    // Need to have a configured catalog for getReplicationOutput
    replicationWorkerHelper.startDestination(
        mock(AirbyteDestination.class),
        new ReplicationInput().withCatalog(new ConfiguredAirbyteCatalog()),
        mock(Path.class));

    when(syncStatsTracker.getTotalBytesEmitted()).thenReturn(100L);
    when(syncStatsTracker.getTotalRecordsEmitted()).thenReturn(10L);
    when(syncStatsTracker.getTotalBytesCommitted()).thenReturn(50L);
    when(syncStatsTracker.getTotalRecordsCommitted()).thenReturn(5L);

    final var summary = replicationWorkerHelper.getReplicationOutput();
    assertEquals(50L, summary.getReplicationAttemptSummary().getBytesSynced());
    assertEquals(5L, summary.getReplicationAttemptSummary().getRecordsSynced());
  }

  @Test
  void testAnalyticsMessageHandling() throws IOException {
    mockSupportRefreshes(false);
    // Need to pass in a replication context
    final ConfiguredAirbyteCatalog catalog = mock(ConfiguredAirbyteCatalog.class);
    when(destinationCatalogGenerator.generateDestinationCatalog(any()))
        .thenReturn(new DestinationCatalogGenerator.CatalogGenerationResult(catalog, Map.of()));
    replicationWorkerHelper.initialize(
        replicationContext,
        mock(ReplicationFeatureFlags.class),
        mock(Path.class),
        catalog,
        mock(State.class));
    // Need to have a configured catalog for getReplicationOutput
    replicationWorkerHelper.startDestination(
        mock(AirbyteDestination.class),
        new ReplicationInput().withCatalog(new ConfiguredAirbyteCatalog()),
        mock(Path.class));

    replicationWorkerHelper.startSource(
        mock(AirbyteSource.class),
        new ReplicationInput().withCatalog(new ConfiguredAirbyteCatalog()),
        mock(Path.class));

    final AirbyteMessage sourceMessage = new AirbyteMessage().withType(Type.TRACE).withTrace(new AirbyteTraceMessage()
        .withType(AirbyteTraceMessage.Type.ANALYTICS).withAnalytics(new AirbyteAnalyticsTraceMessage().withType("from").withValue("source")));
    final AirbyteMessage logMessage = new AirbyteMessage().withType(Type.LOG).withLog(new AirbyteLogMessage().withMessage("test"));
    final AirbyteMessage destinationMessage = new AirbyteMessage().withType(Type.TRACE).withTrace(new AirbyteTraceMessage()
        .withType(AirbyteTraceMessage.Type.ANALYTICS).withAnalytics(new AirbyteAnalyticsTraceMessage().withType("from").withValue("destination")));
    when(mapper.mapMessage(any())).thenAnswer(i -> i.getArgument(0));
    when(mapper.revertMap(any())).thenAnswer(i -> i.getArgument(0));

    replicationWorkerHelper.processMessageFromSource(sourceMessage);
    // this shouldn't be passed to the message tracker
    replicationWorkerHelper.processMessageFromSource(logMessage);
    replicationWorkerHelper.processMessageFromDestination(destinationMessage);

    replicationWorkerHelper.endOfReplication();
    verify(analyticsMessageTracker, times(1)).addMessage(sourceMessage, AirbyteMessageOrigin.SOURCE);
    verify(analyticsMessageTracker, times(1)).addMessage(destinationMessage, AirbyteMessageOrigin.DESTINATION);
    verify(analyticsMessageTracker, times(1)).flush();
  }

  @Test
  void testMessageIsMappedAfterProcessing() {
    final AirbyteMessage sourceRawMessage = mock(AirbyteMessage.class);
    final AirbyteMessage mappedSourceMessage = mock(AirbyteMessage.class);

    doReturn(sourceRawMessage).when(replicationWorkerHelper).internalProcessMessageFromSource(sourceRawMessage);
    when(mapper.mapMessage(sourceRawMessage)).thenReturn(mappedSourceMessage);

    final Optional<AirbyteMessage> processedMessageFromSource = replicationWorkerHelper.processMessageFromSource(sourceRawMessage);

    assertEquals(Optional.of(mappedSourceMessage), processedMessageFromSource);
  }

  @Test
  void testMessageMapIsRevertedBeforeProcessing() {
    final AirbyteMessage destinationRawMessage = mock(AirbyteMessage.class);
    final AirbyteMessage mapRevertedDestinationMessage = mock(AirbyteMessage.class);

    when(mapper.revertMap(destinationRawMessage)).thenReturn(mapRevertedDestinationMessage);
    doNothing().when(replicationWorkerHelper).internalProcessMessageFromDestination(mapRevertedDestinationMessage);

    replicationWorkerHelper.processMessageFromDestination(destinationRawMessage);

    verify(replicationWorkerHelper, times(1)).internalProcessMessageFromDestination(mapRevertedDestinationMessage);
  }

  @Test
  void callsStreamStatusTrackerOnSourceMessage() throws IOException {
    mockSupportRefreshes(true);
    final ConfiguredAirbyteCatalog catalog = mock(ConfiguredAirbyteCatalog.class);
    when(destinationCatalogGenerator.generateDestinationCatalog(any()))
        .thenReturn(new DestinationCatalogGenerator.CatalogGenerationResult(catalog, Map.of()));

    replicationWorkerHelper.initialize(
        replicationContext,
        mock(ReplicationFeatureFlags.class),
        mock(Path.class),
        catalog,
        mock(State.class));

    final AirbyteMessage message = mock(AirbyteMessage.class);

    replicationWorkerHelper.processMessageFromSource(message);

    verify(streamStatusTracker, times(1)).track(message);
  }

  @Test
  void callsStreamStatusTrackerOnDestinationMessage() throws IOException {
    mockSupportRefreshes(true);
    final ConfiguredAirbyteCatalog catalog = mock(ConfiguredAirbyteCatalog.class);
    when(destinationCatalogGenerator.generateDestinationCatalog(any()))
        .thenReturn(new DestinationCatalogGenerator.CatalogGenerationResult(catalog, Map.of()));
    replicationWorkerHelper.initialize(
        replicationContext,
        mock(ReplicationFeatureFlags.class),
        mock(Path.class),
        catalog,
        mock(State.class));

    final AirbyteMessage message = mock(AirbyteMessage.class);
    when(mapper.revertMap(message)).thenReturn(message);

    replicationWorkerHelper.processMessageFromDestination(message);

    verify(streamStatusTracker, times(1)).track(message);
  }

  @ValueSource(booleans = {true, false})
  @ParameterizedTest
  void testSupportRefreshesIsPassed(final boolean supportRefreshes) throws Exception {
    mockSupportRefreshes(supportRefreshes);
    // Need to pass in a replication context
    final ConfiguredAirbyteCatalog catalog = buildConfiguredAirbyteCatalog();
    when(destinationCatalogGenerator.generateDestinationCatalog(any()))
        .thenReturn(new DestinationCatalogGenerator.CatalogGenerationResult(catalog, Map.of()));
    replicationWorkerHelper.initialize(
        replicationContext,
        mock(ReplicationFeatureFlags.class),
        mock(Path.class),
        catalog,
        mock(State.class));

    final ArgumentCaptor<WorkerDestinationConfig> configCaptor = ArgumentCaptor.forClass(WorkerDestinationConfig.class);
    final AirbyteDestination destination = mock(AirbyteDestination.class);

    final ReplicationInput input = new ReplicationInput().withCatalog(new ConfiguredAirbyteCatalog());
    replicationWorkerHelper.startDestination(destination, input, mock(Path.class));

    verify(destination).start(configCaptor.capture(), any());
    assertEquals(supportRefreshes, configCaptor.getValue().getSupportRefreshes());
  }

  @Test
  void testApplyTransformationNoMapper() throws IOException {
    mockSupportRefreshes(false);
    final ConfiguredAirbyteCatalog catalog = mock(ConfiguredAirbyteCatalog.class);
    when(destinationCatalogGenerator.generateDestinationCatalog(any()))
        .thenReturn(new DestinationCatalogGenerator.CatalogGenerationResult(catalog, Map.of()));
    // Need to pass in a replication context
    replicationWorkerHelper.initialize(
        replicationContext,
        mock(ReplicationFeatureFlags.class),
        mock(Path.class),
        catalog,
        mock(State.class));

    final AirbyteMessage recordMessage = new AirbyteMessage().withType(Type.RECORD)
        .withRecord(new AirbyteRecordMessage().withStream("stream").withData(Jsons.jsonNode(Map.of("column", "value"))));
    final AirbyteMessage copiedRecordMessage = Jsons.clone(recordMessage);

    replicationWorkerHelper.applyTransformationMappers(new AirbyteJsonRecordAdapter(recordMessage));

    assertEquals(copiedRecordMessage, recordMessage);
    verifyNoInteractions(recordMapper);
  }

  @Test
  void testApplyTransformationMapper() throws IOException {
    mockSupportRefreshes(false);

    final ConfiguredAirbyteCatalog catalog = mock(ConfiguredAirbyteCatalog.class);
    final ConfiguredAirbyteStream stream = mock(ConfiguredAirbyteStream.class);
    final List<MapperConfig> mappers = List.of(new MapperConfig() {

      @NotNull
      @Override
      public String name() {
        return "test";
      }

      @Nullable
      @Override
      public String documentationUrl() {
        return null;
      }

      @Nullable
      @Override
      public UUID id() {
        return null;
      }

      @NotNull
      @Override
      public Object config() {
        return Map.of();
      }

    });

    when(stream.getStreamDescriptor()).thenReturn(new StreamDescriptor().withName("stream"));
    when(stream.getMappers()).thenReturn(mappers);
    when(catalog.getStreams()).thenReturn(List.of(stream));
    when(destinationCatalogGenerator.generateDestinationCatalog(any()))
        .thenReturn(new DestinationCatalogGenerator.CatalogGenerationResult(catalog, Map.of()));
    // Need to pass in a replication context
    replicationWorkerHelper.initialize(
        replicationContext,
        mock(ReplicationFeatureFlags.class),
        mock(Path.class),
        catalog,
        mock(State.class));

    final AirbyteMessage recordMessage =
        new AirbyteMessage().withType(Type.RECORD)
            .withRecord(new AirbyteRecordMessage().withStream("stream").withData(Jsons.jsonNode(Map.of("column", "value"))));
    final AirbyteJsonRecordAdapter recordAdapter = new AirbyteJsonRecordAdapter(recordMessage);

    replicationWorkerHelper.applyTransformationMappers(recordAdapter);

    verify(recordMapper).applyMappers(recordAdapter, mappers);
  }

  private void mockSupportRefreshes(final boolean supportsRefreshes) throws IOException {
    when(actorDefinitionVersionApi.resolveActorDefinitionVersionByTag(any())).thenReturn(
        new ResolveActorDefinitionVersionResponse(
            UUID.randomUUID(),
            "dockerRepository",
            "dockerImageTag",
            supportsRefreshes,
            false));
  }

  private ConfiguredAirbyteCatalog buildConfiguredAirbyteCatalog() {
    return new ConfiguredAirbyteCatalog();
  }

}

File: airbyte-container-orchestrator/src/test/kotlin/io/airbyte/container/orchestrator/worker/ReplicationWorkerHelperTest.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container.orchestrator.worker

import com.fasterxml.jackson.databind.ObjectMapper
import io.airbyte.config.ConfiguredAirbyteCatalog
import io.airbyte.config.ConfiguredAirbyteStream
import io.airbyte.config.MapperConfig
import io.airbyte.config.StandardSyncSummary.ReplicationStatus
import io.airbyte.config.StreamDescriptor
import io.airbyte.config.adapters.AirbyteJsonRecordAdapter
import io.airbyte.container.orchestrator.bookkeeping.AirbyteMessageOrigin
import io.airbyte.container.orchestrator.bookkeeping.AirbyteMessageTracker
import io.airbyte.container.orchestrator.bookkeeping.SyncStatsTracker
import io.airbyte.container.orchestrator.bookkeeping.events.ReplicationAirbyteMessageEventPublishingHelper
import io.airbyte.container.orchestrator.bookkeeping.streamstatus.StreamStatusTracker
import io.airbyte.container.orchestrator.tracker.AnalyticsMessageTracker
import io.airbyte.container.orchestrator.tracker.StreamStatusCompletionTracker
import io.airbyte.container.orchestrator.tracker.ThreadedTimeTracker
import io.airbyte.container.orchestrator.worker.filter.FieldSelector
import io.airbyte.mappers.application.RecordMapper
import io.airbyte.mappers.transformations.DestinationCatalogGenerator
import io.airbyte.metrics.MetricClient
import io.airbyte.protocol.models.v0.AirbyteAnalyticsTraceMessage
import io.airbyte.protocol.models.v0.AirbyteMessage
import io.airbyte.protocol.models.v0.AirbyteMessage.Type.LOG
import io.airbyte.protocol.models.v0.AirbyteMessage.Type.RECORD
import io.airbyte.protocol.models.v0.AirbyteMessage.Type.TRACE
import io.airbyte.protocol.models.v0.AirbyteRecordMessage
import io.airbyte.protocol.models.v0.AirbyteTraceMessage
import io.airbyte.protocol.models.v0.AirbyteTraceMessage.Type.ANALYTICS
import io.airbyte.workers.internal.AirbyteMapper
import io.mockk.MockKAnnotations
import io.mockk.Runs
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.impl.annotations.SpyK
import io.mockk.junit5.MockKExtension
import io.mockk.just
import io.mockk.mockk
import io.mockk.spyk
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.util.UUID

@ExtendWith(MockKExtension::class)
class ReplicationWorkerHelperTest {
  @MockK(relaxed = true)
  lateinit var fieldSelector: FieldSelector

  @MockK(relaxed = true)
  lateinit var mapper: AirbyteMapper

  @MockK(relaxed = true)
  lateinit var messageTracker: AirbyteMessageTracker

  @MockK(relaxed = true)
  lateinit var eventPublisher: ReplicationAirbyteMessageEventPublishingHelper

  @MockK(relaxed = true)
  lateinit var timeTracker: ThreadedTimeTracker

  @MockK(relaxed = true)
  lateinit var analyticsTracker: AnalyticsMessageTracker

  @MockK(relaxed = true)
  lateinit var streamStatusCompletionTracker: StreamStatusCompletionTracker

  @MockK(relaxed = true)
  lateinit var streamStatusTracker: StreamStatusTracker

  @MockK(relaxed = true)
  lateinit var recordMapper: RecordMapper

  @MockK(relaxed = true)
  lateinit var replicationWorkerState: ReplicationWorkerState

  @MockK(relaxed = true)
  lateinit var context: ReplicationContextProvider.Context

  @MockK(relaxed = true)
  lateinit var destinationCatalogGenerator: DestinationCatalogGenerator

  @MockK(relaxed = true)
  lateinit var metricClient: MetricClient

  @MockK(relaxed = true)
  lateinit var syncStatsTracker: SyncStatsTracker

  @InjectMockKs
  @SpyK
  lateinit var helper: ReplicationWorkerHelper

  @BeforeEach
  fun setUp() {
    MockKAnnotations.init(this)

    every { context.replicationContext } returns mockk(relaxed = true)
    every { context.supportRefreshes } returns false
    val configuredCatalog = mockk<ConfiguredAirbyteCatalog>(relaxed = true)
    every { context.configuredCatalog } returns configuredCatalog
    every { destinationCatalogGenerator.generateDestinationCatalog(configuredCatalog) } answers {
      DestinationCatalogGenerator.CatalogGenerationResult(configuredCatalog, emptyMap())
    }

    every { syncStatsTracker.getTotalBytesEmitted() } returns 0L
    every { syncStatsTracker.getTotalRecordsEmitted() } returns 0L
  }

  @Test
  fun `getReplicationOutput returns correct stats for completed sync`() {
    every { replicationWorkerState.cancelled } returns false
    every { replicationWorkerState.hasFailed } returns false

    every { syncStatsTracker.getTotalBytesCommitted() } returns 50L
    every { syncStatsTracker.getTotalRecordsCommitted() } returns 5L
    every { syncStatsTracker.getUnreliableStateTimingMetrics() } returns false

    val output = helper.getReplicationOutput()

    assertEquals(50L, output.replicationAttemptSummary.bytesSynced)
    assertEquals(5L, output.replicationAttemptSummary.recordsSynced)
    assertEquals(ReplicationStatus.COMPLETED, output.replicationAttemptSummary.status)
  }

  @Test
  fun `analytics messages are tracked and flushed`() {
    every { replicationWorkerState.cancelled } returns false
    every { replicationWorkerState.hasFailed } returns false
    every { mapper.mapMessage(any()) } answers { firstArg<AirbyteMessage>() }
    every { mapper.revertMap(any()) } answers { firstArg<AirbyteMessage>() }

    val sourceMsg =
      AirbyteMessage()
        .withType(TRACE)
        .withTrace(
          AirbyteTraceMessage()
            .withType(ANALYTICS)
            .withAnalytics(AirbyteAnalyticsTraceMessage().withType("from").withValue("src")),
        )
    val logMsg = AirbyteMessage().withType(LOG)
    val destMsg =
      AirbyteMessage()
        .withType(TRACE)
        .withTrace(
          AirbyteTraceMessage()
            .withType(ANALYTICS)
            .withAnalytics(AirbyteAnalyticsTraceMessage().withType("from").withValue("dest")),
        )

    helper.processMessageFromSource(sourceMsg)
    helper.processMessageFromSource(logMsg)
    helper.processMessageFromDestination(destMsg)
    helper.endOfReplication()

    verify(exactly = 1) { analyticsTracker.addMessage(sourceMsg, AirbyteMessageOrigin.SOURCE) }
    verify(exactly = 1) { analyticsTracker.addMessage(destMsg, AirbyteMessageOrigin.DESTINATION) }
    verify(exactly = 1) { analyticsTracker.flush() }
  }

  @Test
  fun `processMessageFromSource maps after internal processing`() {
    val raw = mockk<AirbyteMessage>(relaxed = true)
    val mapped = mockk<AirbyteMessage>(relaxed = true)
    every { helper.internalProcessMessageFromSource(raw) } returns raw
    every { mapper.mapMessage(raw) } returns mapped

    val result = helper.processMessageFromSource(raw)
    assertTrue(result.isPresent)
    assertEquals(mapped, result.get())
  }

  @Test
  fun `processMessageFromDestination reverts map before processing`() {
    val raw = mockk<AirbyteMessage>(relaxed = true)
    val reverted = mockk<AirbyteMessage>(relaxed = true)
    every { mapper.revertMap(raw) } returns reverted
    every { helper.internalProcessMessageFromDestination(reverted) } just Runs

    helper.processMessageFromDestination(raw)
    verify(exactly = 1) { helper.internalProcessMessageFromDestination(reverted) }
  }

  @Test
  fun `stream status tracker is called on source messages`() {
    val msg = mockk<AirbyteMessage>(relaxed = true)
    helper.internalProcessMessageFromSource(msg)
    verify { streamStatusTracker.track(msg) }
  }

  @Test
  fun `stream status tracker is called on destination messages`() {
    val msg = mockk<AirbyteMessage>(relaxed = true)
    every { mapper.revertMap(msg) } returns msg
    helper.internalProcessMessageFromDestination(msg)
    verify { streamStatusTracker.track(msg) }
  }

  @Test
  fun `applyTransformationMappers with no mappers does nothing`() {
    every { context.configuredCatalog.streams } returns emptyList()
    val adapter =
      AirbyteJsonRecordAdapter(
        AirbyteMessage()
          .withType(RECORD)
          .withRecord(AirbyteRecordMessage().withStream("s").withData(ObjectMapper().createObjectNode())),
      )
    helper.applyTransformationMappers(adapter)
    verify(exactly = 0) { recordMapper.applyMappers(any(), any<List<MapperConfig>>()) }
  }

  @Test
  fun `applyTransformationMappers with configured mappers applies them`() {
    val streamDescriptor =
      StreamDescriptor()
        .withName("s")
    val mapperConfig =
      object : MapperConfig {
        override fun name() = "test"

        override fun documentationUrl(): String? = null

        override fun id(): UUID? = null

        override fun config(): Any = emptyMap<String, String>()
      }
    val configuredStream = mockk<ConfiguredAirbyteStream>(relaxed = true)
    every { configuredStream.streamDescriptor } returns streamDescriptor
    every { configuredStream.mappers } returns listOf(mapperConfig)
    every { context.configuredCatalog.streams } returns listOf(configuredStream)
    every { destinationCatalogGenerator.generateDestinationCatalog(any()) } answers {
      DestinationCatalogGenerator.CatalogGenerationResult(context.configuredCatalog, emptyMap())
    }
    helper =
      spyk(
        ReplicationWorkerHelper(
          fieldSelector,
          mapper,
          messageTracker,
          eventPublisher,
          timeTracker,
          analyticsTracker,
          streamStatusCompletionTracker,
          syncStatsTracker,
          streamStatusTracker,
          recordMapper,
          replicationWorkerState,
          context,
          destinationCatalogGenerator,
          metricClient,
        ),
        recordPrivateCalls = true,
      )

    val adapter =
      AirbyteJsonRecordAdapter(
        AirbyteMessage()
          .withType(RECORD)
          .withRecord(AirbyteRecordMessage().withStream("s").withData(ObjectMapper().createObjectNode())),
      )
    helper.applyTransformationMappers(adapter)

    verify(exactly = 1) { recordMapper.applyMappers(adapter, listOf(mapperConfig)) }
  }
}
